File: app/src/main/java/libertypassage/com/app/basic/Acknowledgement.java
package libertypassage.com.app.basic;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Dialog;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.location.LocationManager;
import android.os.Bundle;
import android.os.Looper;
import android.provider.Settings;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SwitchCompat;
import androidx.core.app.ActivityCompat;

import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.gson.Gson;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Set;

import libertypassage.com.app.R;
import libertypassage.com.app.models.ModelResponse;
import libertypassage.com.app.models.user.ModelUser;
import libertypassage.com.app.utilis.ApiInterface;
import libertypassage.com.app.utilis.ClientInstance;
import libertypassage.com.app.utilis.Constants;
import libertypassage.com.app.utilis.Utility;
import libertypassage.com.app.widgets.CenturyGothicTextview;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;


public class Acknowledgement extends AppCompatActivity implements View.OnClickListener {

    private Context context;
    private final String TAG = Acknowledgement.class.getSimpleName();
    private CenturyGothicTextview tv_next, tv_latLong;
    private ImageView iv_back;
    private String token, address, city, state, country, postalCode, knownName, location = "1",
            macAddress="", wifiIpAddress="";
    private SwitchCompat location_switch, blutooth_switch;
    //  get current location by fused location api
    private FusedLocationProviderClient mFusedLocationClient;
    int PERMISSION_ID = 44;
    double latitudes, longitudes, altitudes;




    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.acknowledgement);

        context = Acknowledgement.this;
        findID();
//        init();
        token = Utility.getSharedPreferences(context, Constants.KEY_BEARER_TOKEN);
    }


    @SuppressLint("MissingPermission")
    private void findID() {
        iv_back = findViewById(R.id.iv_back);
        tv_next = findViewById(R.id.tv_next);
        location_switch = findViewById(R.id.location_switch);
        blutooth_switch = findViewById(R.id.blutooth_switch);
        tv_latLong = findViewById(R.id.tv_latLong);

        iv_back.setOnClickListener(this);
        tv_next.setOnClickListener(this);
        tv_latLong.setVisibility(View.GONE);
        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this);


        location_switch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
                if (b) {
                    location = "1";
                } else {
                    dialogUpdate();
//                    Intent intent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
//                    startActivity(intent);
//                    location_switch.setChecked(true);
                }
            }
        });
    }
    
    
    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.iv_back: {
                finish();
                break;
            }
            case R.id.tv_next: {
                if (location.equals("0")) {
                    dialogUpdate();
                } else {
                    getLastLocation();
                }
                break;
            }
        }
    }


    @SuppressLint("MissingPermission")
    private void getLastLocation() {
        if (checkPermissions()) {
            if (isLocationEnabled()) {
                LocationRequest locationRequest = LocationRequest.create();
                locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
                mFusedLocationClient.getLastLocation().addOnCompleteListener(
                        new OnCompleteListener<Location>() {
                            @Override
                            public void onComplete(@NonNull Task<Location> task) {
                                Location location = task.getResult();
                                if (location == null) {
                                    requestNewLocationData();
                                } else {
                                    tv_latLong.setText("LAT : " + location.getLatitude() + ", LONG : " + location.getLongitude() + "");
                                    latitudes = location.getLatitude();
                                    longitudes = location.getLongitude();
                                    altitudes = location.getAltitude();
                                    getAddress();
                                }
                            }
                        }
                );
            } else {
                Toast.makeText(this, "Turn on location", Toast.LENGTH_LONG).show();
                Intent intent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
                startActivity(intent);
            }
        } else {
            requestPermissions();
        }
    }

    @SuppressLint("MissingPermission")
    private void requestNewLocationData() {
        LocationRequest mLocationRequest = new LocationRequest();
        mLocationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
        mLocationRequest.setInterval(0);
        mLocationRequest.setFastestInterval(0);
        mLocationRequest.setNumUpdates(1);

        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this);
        mFusedLocationClient.requestLocationUpdates(
                mLocationRequest, mLocationCallback,
                Looper.myLooper()
        );

    }

    private LocationCallback mLocationCallback = new LocationCallback() {
        @Override
        public void onLocationResult(LocationResult locationResult) {
            Location mLastLocation = locationResult.getLastLocation();
            tv_latLong.setText("LAT : " + mLastLocation.getLatitude() + ", LONG : " + mLastLocation.getLongitude() + "");
        }
    };

    private boolean checkPermissions() {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED &&
                ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
            return true;
        }
        return false;
    }

    private void requestPermissions() {
        ActivityCompat.requestPermissions(
                this,
                new String[]{Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION},
                PERMISSION_ID
        );
    }

    private boolean isLocationEnabled() {
        LocationManager locationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
        return locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER) || locationManager.isProviderEnabled(
                LocationManager.NETWORK_PROVIDER
        );
    }


    private void getAddress() {
        Geocoder geocoder;
        List<Address> addresses;
        geocoder = new Geocoder(context, Locale.getDefault());
        try {
            addresses = geocoder.getFromLocation(latitudes, longitudes, 1); // Here 1 represent max location result to returned, by documents it recommended 1 to 5
            address = addresses.get(0).getAddressLine(0); // If any additional address line present than only, check with max available address lines by getMaxAddressLineIndex()
            city = addresses.get(0).getLocality();
            state = addresses.get(0).getAdminArea();
            country = addresses.get(0).getCountryName();
            postalCode = addresses.get(0).getPostalCode();
            knownName = addresses.get(0).getFeatureName(); // Only if available else return NULL

            if (Utility.isConnectingToInternet(context)) {
                addTrackUserLocation();
            } else {
                Toast.makeText(context, "Please connect the internet", Toast.LENGTH_LONG).show();
            }

            Log.e("address", address + ", " + city + "," + state + "," + country + postalCode);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == PERMISSION_ID) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                getLastLocation();
            }
        }
    }

    private void addTrackUserLocation() {
        Utility.showProgressDialog(context);
        ApiInterface apiInterface = ClientInstance.getRetrofitInstance().create(ApiInterface.class);
        Call<ModelUser> call = apiInterface.addTrackUserLocation(Constants.KEY_HEADER + token,
                Constants.KEY_BOT, latitudes, longitudes, altitudes, macAddress, wifiIpAddress);

        call.enqueue(new Callback<ModelUser>() {
            @Override
            public void onResponse(Call<ModelUser> call, Response<ModelUser> response) {
                Utility.stopProgressDialog(context);
                ModelUser model = response.body();

                if (model != null && model.getError().equals(false)) {

                    Utility.setSharedPreference(context, Constants.KEY_START, "3");
                    Intent intent = new Intent(context, MyTemperature.class);
                    startActivity(intent);

                    
                } else if (model != null && model.getError().equals(true)) {
                    Utility.stopProgressDialog(context);
                    Toast.makeText(context, model.getMessage(), Toast.LENGTH_LONG).show();
                }
            }

            @Override
            public void onFailure(Call<ModelUser> call, Throwable t) {
                Utility.stopProgressDialog(context);
                Toast.makeText(context, Constants.ERROR_MSG, Toast.LENGTH_LONG).show();
            }
        });
    }

    private void dialogUpdate() {
        Dialog dialog;
        dialog = new Dialog(Acknowledgement.this);
        dialog.setContentView(R.layout.dialog_location_required);
        dialog.getWindow().setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        dialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        dialog.setCancelable(false);

        ImageView cancel = dialog.findViewById(R.id.cancel);
        CenturyGothicTextview tv_notAllow = dialog.findViewById(R.id.tv_notAllow);
        CenturyGothicTextview tv_ok = dialog.findViewById(R.id.tv_ok);

        cancel.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                location = "0";
                dialog.dismiss();
            }
        });
        tv_notAllow.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                location_switch.setChecked(false);
                location = "0";
                dialog.dismiss();
            }
        });
        tv_ok.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                location_switch.setChecked(true);
                location = "1";
                dialog.dismiss();
            }
        });

        dialog.show();
    }

}

File: app/src/main/java/libertypassage/com/corporate/view/Acknowledgement.kt
package libertypassage.com.corporate.view

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.content.IntentSender
import android.content.pm.PackageManager
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.location.LocationManager
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.net.Uri
import android.net.wifi.WifiManager
import android.os.Build
import android.os.Bundle
import android.os.Looper
import android.provider.Settings
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.location.*
import kotlinx.android.synthetic.main.acknowledgement.*
import kotlinx.android.synthetic.main.activity_header.*
import libertypassage.com.corporate.R
import libertypassage.com.corporate.model.ModelTrackUserLocation
import libertypassage.com.corporate.retofit.ApiInterface
import libertypassage.com.corporate.retofit.ClientInstance
import libertypassage.com.corporate.services.GPSTrackerLocation
import libertypassage.com.corporate.utilities.Constants
import libertypassage.com.corporate.utilities.DialogProgress
import libertypassage.com.corporate.utilities.Utility
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*


class Acknowledgement : AppCompatActivity(), View.OnClickListener {
    private var TAG = Acknowledgement::class.java.simpleName
    lateinit var context: Context
    var dialogProgress: DialogProgress? = null
    var mFusedLocationClient: FusedLocationProviderClient? = null
    private var PERMISSION_ID = 44
    private var latitudes = 0.0
    private var longitudes = 0.0
    private var altitudes = 0.0
    private var token = ""
    private var wifiMacAddress = ""
    private var wifiIpAddress = ""
    private var wifiMode = ""
    private var corporationId = ""
    private var location = ""

    @RequiresApi(Build.VERSION_CODES.Q)
    val permissions = arrayOf<String?>(
        Manifest.permission.ACCESS_COARSE_LOCATION,
        Manifest.permission.ACCESS_FINE_LOCATION,
        Manifest.permission.ACCESS_BACKGROUND_LOCATION
    )




    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.acknowledgement)
        context = this@Acknowledgement

        token = Utility.getSharedPreferences(context, Constants.KEY_BEARER_TOKEN)!!
        corporationId = Utility.getSharedPreferences(context, Constants.KEY_CORPORATION_ID)!!
        dialogProgress = DialogProgress(context)
        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this)

        iv_back.setOnClickListener(this)
        tv_next.setOnClickListener(this)

        location_switch.setOnCheckedChangeListener { compoundButton, b ->
            if (b) {
                location = "1"
            } else {
                dialogUpdate()
            }
        }
    }


    @RequiresApi(Build.VERSION_CODES.Q)
    override fun onClick(v: View?) {
        when (v!!.id) {
            R.id.iv_back -> {
                finish()
            }

            R.id.tv_next -> {
                if (location == "0") {
                    dialogUpdate()
                } else {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        requestPermissions(permissions, 456)
                    } else {
                        if (isLocationEnabled()) {
                            Utility.getMyLocation(context)
                        } else {
                            dialogAlwaysLocationOn()
                        }
                        getGPSLocation()
                    }
                }
            }
        }
    }

    private fun getGPSLocation() {
        val gps = GPSTrackerLocation(context)
        val manager = getSystemService(LOCATION_SERVICE) as LocationManager
        if (!manager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
            enableGPSSettingsRequest(this@Acknowledgement)
        } else {
            latitudes = gps.getLatitude()
            longitudes = gps.getLongitude()
            altitudes = gps.getAltitude()

            Utility.setSharedPreference(context, Constants.KEY_LAT, latitudes.toString())
            Utility.setSharedPreference(context, Constants.KEY_LONG, longitudes.toString())
            Utility.setSharedPreference(context, Constants.KEY_ALT, altitudes.toString())
            getWifiDetails()
        }
    }

    @RequiresApi(Build.VERSION_CODES.Q)
    @SuppressLint("MissingPermission")
    private fun getLastLocation() {
        if (checkPermissions()) {
            if (isLocationEnabled()) {
                val locationRequest = LocationRequest.create()
                locationRequest.priority = LocationRequest.PRIORITY_HIGH_ACCURACY
                mFusedLocationClient!!.lastLocation.addOnCompleteListener { task ->
                    val location = task.result
                    if (location == null) {
                        requestNewLocationData()
                    } else {
                        latitudes = location.latitude
                        longitudes = location.longitude
                        altitudes = location.altitude
                        getWifiDetails()
                    }
                }
            } else {
                Toast.makeText(this, "Turn on location", Toast.LENGTH_LONG).show()
                val intent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
                startActivity(intent)
            }
        } else {
            requestPermissions()
        }
    }

    @SuppressLint("MissingPermission")
    private fun requestNewLocationData() {
        val mLocationRequest = LocationRequest()
        mLocationRequest.priority = LocationRequest.PRIORITY_HIGH_ACCURACY
        mLocationRequest.interval = 0
        mLocationRequest.fastestInterval = 0
        mLocationRequest.numUpdates = 1
        mFusedLocationClient?.requestLocationUpdates(
            mLocationRequest, mLocationCallback,
            Looper.myLooper()
        )
    }

    private val mLocationCallback: LocationCallback = object : LocationCallback() {
        override fun onLocationResult(locationResult: LocationResult) {
//            val mLastLocation = locationResult.lastLocation
        }
    }

    @RequiresApi(Build.VERSION_CODES.Q)
    private fun checkPermissions(): Boolean {
        return ActivityCompat.checkSelfPermission(this,
            Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED &&
                ActivityCompat.checkSelfPermission(this,
                    Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED &&
                ActivityCompat.checkSelfPermission(this,
                    Manifest.permission.ACCESS_BACKGROUND_LOCATION) == PackageManager.PERMISSION_GRANTED
    }

    @RequiresApi(Build.VERSION_CODES.Q)
    private fun requestPermissions() {
        ActivityCompat.requestPermissions(
            this,
            arrayOf(
                Manifest.permission.ACCESS_COARSE_LOCATION,
                Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_BACKGROUND_LOCATION
            ),
            PERMISSION_ID
        )
    }

    private fun isLocationEnabled(): Boolean {
        val locationManager = getSystemService(LOCATION_SERVICE) as LocationManager
        return locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER) || locationManager.isProviderEnabled(
            LocationManager.NETWORK_PROVIDER
        )
    }

    @RequiresApi(Build.VERSION_CODES.Q)
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
//        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
//        if (requestCode == PERMISSION_ID) {
//            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
//                getLastLocation()
//            }
//        }
        when (requestCode) {
        456 -> {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                if (isLocationEnabled()) {
                    Utility.getMyLocation(context)
                } else {
                    dialogAlwaysLocationOn()
                }
                getGPSLocation()
                // permission was granted, yay! do the
            } else {
                // permission denied, boo! Disable the
                dialogRequestPermission()
            }
            return
        }
        }
    }

    private fun dialogUpdate() {
        val dialog = Dialog(context)
        dialog.setContentView(R.layout.dialog_location_required)
        dialog.window!!.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        dialog.window!!.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        dialog.setCancelable(false)

        val ivCancel = dialog.findViewById(R.id.ivCancel) as ImageView
        val tvDontAllow = dialog.findViewById(R.id.tvDontAllow) as TextView
        val tvOk = dialog.findViewById(R.id.tvOk) as TextView

        ivCancel.setOnClickListener {
            location = "0"
            dialog.dismiss()
        }

        tvDontAllow.setOnClickListener {
            location_switch.isChecked = false
            location = "0"
            dialog.dismiss()
        }

        tvOk.setOnClickListener {
            location_switch.isChecked = true
            location = "1"
            dialog.dismiss()
        }
        dialog.show()
    }

    private fun getWifiDetails() {
        //check network type
        val connManager = getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
        val mWifi = connManager.getNetworkInfo(ConnectivityManager.TYPE_WIFI)
        val wifiManager =
            super.getApplicationContext().getSystemService(WIFI_SERVICE) as WifiManager
        val wInfo = wifiManager.connectionInfo
        if(wifiManager.isWifiEnabled){
            if(Objects.requireNonNull<NetworkInfo>(mWifi).isConnected) {
                wifiIpAddress = wInfo.ssid
                wifiMacAddress = wInfo.bssid
                wifiMode = "Connected"
            }else{
                val wifiList = wifiManager.scanResults
                val ssidList = ArrayList<String>()
                val bssidList = ArrayList<String>()
                for (scanResult in wifiList) {
                    ssidList.add(scanResult.SSID)
                    bssidList.add(scanResult.BSSID)
                }
                if (ssidList.size != 0) {
                    wifiIpAddress = ssidList[0]
                    wifiMacAddress = bssidList[0]
                    wifiMode = "Searchable"
                }
            }
        }

        if (Utility.isConnectingToInternet(context)) {
            if (latitudes != 0.0) {
                trackUserLocation()
            }
        } else {
            Toast.makeText(context, "Please connect the internet", Toast.LENGTH_LONG).show()
        }
    }

    private fun trackUserLocation() {
        Log.e("LocationUpdateServer", "Continue")
        val apiInterface: ApiInterface = ClientInstance.retrofitInstance!!.create(ApiInterface::class.java)
        val call: Call<ModelTrackUserLocation> = apiInterface.trackUserLocation(Constants.KEY_HEADER + token,
            Constants.KEY_BOT, latitudes, longitudes, altitudes, wifiMacAddress, "$wifiIpAddress-$wifiMode", "0")

        call.enqueue(object : Callback<ModelTrackUserLocation?> {
            override fun onResponse(call: Call<ModelTrackUserLocation?>, response: Response<ModelTrackUserLocation?>) {
                val model: ModelTrackUserLocation? = response.body()

                if (model != null && model.error?.equals(false)!!) {
                    Utility.setSharedPreference(context, Constants.KEY_START, "3")
                    val intent = Intent(context, MyTemperature::class.java)
                    startActivity(intent)

                } else if (model != null && model.error?.equals(true)!!) {
                    Log.e("ErrorResponse", model.message!!)
                }
            }

            override fun onFailure(call: Call<ModelTrackUserLocation?>, t: Throwable) {
                Log.e("model", "onFailure    " + t.message)
            }
        })
    }

    private fun enableGPSSettingsRequest(context: Context) {
        val googleApiClient = GoogleApiClient.Builder(context).addApi(LocationServices.API).build()
        googleApiClient.connect()
        val locationRequest = LocationRequest.create()
        locationRequest.priority = LocationRequest.PRIORITY_HIGH_ACCURACY
        val builder = LocationSettingsRequest.Builder().addLocationRequest(locationRequest)
        builder.setAlwaysShow(true)
        val result = LocationServices.SettingsApi.checkLocationSettings(
            googleApiClient,
            builder.build()
        )
        result.setResultCallback { result ->
            val status = result.status
            when (status.statusCode) {
                LocationSettingsStatusCodes.SUCCESS -> Log.e(
                    TAG,
                    "All location settings are satisfied."
                )
                LocationSettingsStatusCodes.RESOLUTION_REQUIRED -> try {
                    status.startResolutionForResult(this@Acknowledgement, ENABLED_GPS_REQUEST)
                } catch (e: IntentSender.SendIntentException) {
                    Log.e(TAG, "PendingIntent unable to execute request.")
                }
                LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE -> Log.e(
                    TAG,
                    "Location settings are inadequate, and cannot be fixed here. Dialog not created."
                )
            }
        }
    }

    private val LOCATION_SETTINGS_REQUEST = 10
    private val ENABLED_GPS_REQUEST = 20
    @RequiresApi(Build.VERSION_CODES.Q)
    @SuppressLint("MissingSuperCall")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            LOCATION_SETTINGS_REQUEST -> if (isLocationEnabled()) {
                getLastLocation()
            } else {
                getGPSLocation()
            }
            ENABLED_GPS_REQUEST -> when (resultCode) {
                RESULT_OK -> getGPSLocation()
                RESULT_CANCELED -> {
                    if (checkPermissions()) {
                        if (isLocationEnabled()) {
                            getLastLocation()
                        } else {
                            getGPSLocation()
                        }
                    } else {
                        requestPermissions()
                    }
                    Log.e("Cancel", "Not enabled GPS")
                }
                else -> {
                }
            }
            104 ->
                if (resultCode != Activity.RESULT_OK) {
                    if (isLocationEnabled()) {
                        Utility.getMyLocation(context)
                    } else {
                        dialogAlwaysLocationOn()
                    }
                    getGPSLocation()
                }
        }
    }

    private fun dialogAlwaysLocationOn() {
        val dialog = Dialog(context)
        dialog.setContentView(R.layout.dialog_location_alwayson)
        dialog.window!!.setLayout(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        dialog.window!!.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        dialog.setCancelable(false)
        val tvOk = dialog.findViewById(R.id.tvOk) as TextView

        tvOk.setOnClickListener {
            startActivityForResult(Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS), 1)
            dialog.dismiss()
        }
        dialog.show()
    }

    private fun dialogRequestPermission() {
        val dialog = Dialog(context)
        dialog.window!!.setLayout(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        dialog.window!!.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        dialog.setCancelable(true)
        dialog.setContentView(R.layout.dialog_runtime_permission)

        val tvCancel = dialog.findViewById(R.id.tvCancel) as TextView
        val tvSetting = dialog.findViewById(R.id.tvSetting) as TextView

        tvCancel.setOnClickListener {
            dialog.dismiss()
        }

        tvSetting.setOnClickListener {
            dialog.dismiss()
            val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
            val uri: Uri = Uri.fromParts("package", packageName, null)
            intent.data = uri
            startActivityForResult(intent, 456)
        }
        dialog.show()
    }

    override fun onBackPressed() {
        super.onBackPressed()
        finish()
    }
}
